# 367. Valid Perfect Square

## The problem:

Given a positive integer num, write a function which returns True if num is a perfect square else False.

Note: Do not use any built-in library function such as sqrt.

Example 1:

Input: 16

Returns: True

Example 2:

Input: 14

Returns: False

## Implementation

Let us begin by copying over the solution class

In [1]:
class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        

And creating the test class:

It is important to test values which are close to large known squares as rounding can be an issue

In [2]:
from nose.tools import assert_equal


class TestIsPerfectSquare(object):

    def test_isPerfectSquare(self):
        solution = Solution()
        
        assert_equal(solution.isPerfectSquare(0), False)
        assert_equal(solution.isPerfectSquare(1), True)
        assert_equal(solution.isPerfectSquare(-4), False)
        assert_equal(solution.isPerfectSquare(1522756), True)
        assert_equal(solution.isPerfectSquare(1522757), False)
        print('Success: test_isPerfectSquare')


def main():
    test = TestIsPerfectSquare()
    test.test_isPerfectSquare()



What do we know about square numbers?

1. They are always positive
2. The square root of a square is in integer

We cannot implement sqrt from math here, we need to use another fact.
$\sqrt(x) = x^\frac{1}{2}$

In [3]:
print(16 ** 0.5)
print(17 ** 0.5)


4.0
4.123105625617661


Whilst returned as floats, it is clear that 4.0 is an integer whilst 4.123... is not.


Let's have a look at much larger numbers:


In [4]:
print(1522756**0.5)
print(1522757**0.5)


1234.0
1234.0004051863193


The square root of 1522757 is not an integer, this is good.

So, if the square root of the number is an integer then the number is a perfect square, if not then it isn't.

## Pseudocode:

1. raise a number to a power of 0.5

2. If number is an integer, return True, else False

For item 2, we can divide the returned number by a rounded version of that number.
If it isn't 1, then it is not a square number.

In [5]:
num = 1522756
res = num**0.5
res / int(res) == 1

True

In [6]:
num = 1522757
res = num**0.5
res / int(res) == 1

False

The problem appears to be solved, but wait, what if a negative number is passed?

In [7]:
num = -4
res = num**0.5
res / int(res) == 1

TypeError: can't convert complex to int

The error is raised because the square root of a negative number is complex.
We can split the complex and real parts if negative numbers need to be dealt with, however, in our case we need to filter these out:

In [8]:
num = -4
if num <= 0:
    print("Invalid input")
else:
    res = num**0.5
    res / int(res) == 1


Invalid input


Excellent, now to write the code into the function:

In [9]:
class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        if num <= 0:
            return False
        res = num**0.5
        return res / int(res) == 1

and running it

In [10]:
main()

Success: test_isPerfectSquare


--------------------------

Solution has been successfully implemented

--------------------------